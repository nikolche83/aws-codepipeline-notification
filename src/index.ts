import { CodePipelineExecutionStateChangeDetectionEventRule } from '@gammarers/aws-codesuite-state-change-detection-event-rules';
import { ResourceAutoNaming, ResourceDefaultNaming, ResourceNaming, ResourceNamingType } from '@gammarers/aws-resource-naming';
import { Duration, Names, Stack, StackProps } from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import { Construct } from 'constructs';
import { NotificationStateMachine } from './resources/notification-state-machine';

export { ResourceAutoNaming, ResourceDefaultNaming, ResourceNamingType as CodePipelineExecutionStateChangeNotificationStackResourceNamingType };

export interface TargetResourceProperty {
  readonly tagKey: string;
  readonly tagValues: string[];
}

export interface NotificationsProperty {
  readonly emails?: string[];
}
export interface CodePipelineExecutionStateChangeNotificationStackProps extends StackProps {
  readonly targetResource: TargetResourceProperty;
  readonly enabled?: boolean;
  readonly notifications: NotificationsProperty;
  readonly resourceNamingOption?: ResourceNamingOption;
}

export interface CustomNaming {
  readonly type: ResourceNamingType.CUSTOM;
  readonly stateMachineName: string;
  readonly notificationTopicName: string;
  readonly notificationTopicDisplayName: string;
  readonly pipelineEventDetectionRuleName: string;
}

export type ResourceNamingOption = ResourceDefaultNaming | ResourceAutoNaming | CustomNaming;

export class CodePipelineExecutionStateChangeNotificationStack extends Stack {
  constructor(scope: Construct, id: string, props: CodePipelineExecutionStateChangeNotificationStackProps) {
    super(scope, id, props);

    // ðŸ‘‡ Create random 8 length string
    const random = ResourceNaming.createRandomString(`${Names.uniqueId(scope)}.${Names.uniqueId(this)}`);
    // ðŸ‘‡ Auto naeming
    const autoNaming = {
      stateMachineName: `codepipeline-exec-state-change-notification-${random}-machine`,
      notificationTopicName: `codepipeline-execution-state-change-notification-${random}-topic`,
      notificationTopicDisplayName: 'CodePipeline Execution state change Notification Topic',
      pipelineEventDetectionRuleName: `codepipeline-exe-state-change-${random}-detection-event-rule`,
    };
    // ðŸ‘‡ Resource Names
    const names = ResourceNaming.naming(autoNaming, props.resourceNamingOption as ResourceNaming.ResourceNamingOption);

    // ðŸ‘‡ SNS Topic for notifications
    const topic: sns.Topic = new sns.Topic(this, 'NotificationTopic', {
      topicName: names.notificationTopicName,
      displayName: names.notificationTopicDisplayName,
    });

    //    const secret = cdk.SecretValue.secretsManager('my-email-array-secret');
    //    const emails = JSON.parse(secret.unsafeUnwrap()) as string[];

    // Subscribe an email endpoint to the topic
    const emails = props.notifications.emails ?? [];
    for (const email of emails) {
      topic.addSubscription(new subscriptions.EmailSubscription(email));
    }

    // Subscribe a HTTP endpoint (Slack Webhook) to the topic
    // topic.addSubscription(new subs.UrlSubscription('https://hooks.slack.com/services/YOUR/SLACK/WEBHOOK'));

    // ðŸ‘‡ Create State Machine
    const stateMachine = new NotificationStateMachine(this, 'StateMachine', {
      stateMachineName: names.stateMachineName,
      timeout: Duration.minutes(5),
      notificationTopic: topic,
    });

    // ðŸ‘‡ Rule state
    //enabled
    const enableRule: boolean = (() => {
      return props?.enabled === undefined || props.enabled;
    })();

    // ðŸ‘‡ Create EventBridge Rule
    new CodePipelineExecutionStateChangeDetectionEventRule(this, 'EventRule', {
      ruleName: names.pipelineEventDetectionRuleName,
      enabled: enableRule,
      targets: [
        new targets.SfnStateMachine(stateMachine, {
          input: events.RuleTargetInput.fromObject({
            event: events.EventField.fromPath('$'),
            params: {
              tagKey: props.targetResource.tagKey,
              tagValues: props.targetResource.tagValues,
            },
          }),
        }),
      ],
    });
  }
}
