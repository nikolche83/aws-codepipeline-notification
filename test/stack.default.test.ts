import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CodePipelineExecutionStateChangeNotificationStack } from '../src';

describe('Stack Default Testing', () => {

  const app = new App();

  const stack = new CodePipelineExecutionStateChangeNotificationStack(app, 'CodePipelineEventNotificationStack', {
    targetResource: {
      tagKey: 'DeployNotification',
      tagValues: ['YES'],
    },
    notifications: {},
  });

  const template = Template.fromStack(stack);

  it('Should match state machine count', () => {
    template.resourceCountIs('AWS::StepFunctions::StateMachine', 1);
  });

  it('Should match iam role count', () => {
    template.resourceCountIs('AWS::IAM::Role', 2);
  });

  it('Should match event rule count', () => {
    template.resourceCountIs('AWS::Events::Rule', 1);
  });

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });

});
