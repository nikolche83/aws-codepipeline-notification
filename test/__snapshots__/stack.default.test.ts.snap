// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Default Testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "EventRule5A491D2C": {
      "Properties": {
        "EventPattern": {
          "detail-type": [
            "CodePipeline Pipeline Execution State Change",
          ],
          "source": [
            "aws.codepipeline",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "StateMachine2E01A3A5",
            },
            "Id": "Target0",
            "InputTransformer": {
              "InputPathsMap": {
                "f1": "$",
              },
              "InputTemplate": "{"event":<f1>,"params":{"tagKey":"DeployNotification","tagValues":["YES"]}}",
            },
            "RoleArn": {
              "Fn::GetAtt": [
                "StateMachineEventsRoleDBCDECD1",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "NotificationTopicEB7A0DF1": {
      "Type": "AWS::SNS::Topic",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"InitPipelineStateEmojiDefinition","States":{"InitPipelineStateEmojiDefinition":{"Type":"Pass","Result":[{"name":"STARTED","emoji":"ü•≥"},{"name":"SUCCEEDED","emoji":"ü§©"},{"name":"RESUMED","emoji":"ü§î"},{"name":"FAILED","emoji":"üò´"},{"name":"STOPPING","emoji":"üòÆ"},{"name":"STOPPED","emoji":"üòå"},{"name":"SUPERSEDED","emoji":"üßê"}],"ResultPath":"$.Definition.StateEmojis","Next":"GetPipeline"},"GetPipeline":{"Next":"GetResourceTagMappingList","Type":"Task","ResultPath":"$.Result.Pipeline","ResultSelector":{"Arn.$":"$.Metadata.PipelineArn"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:codepipeline:getPipeline","Parameters":{"Name.$":"$.event.detail.pipeline"}},"GetResourceTagMappingList":{"Next":"CheckTagFilterArnsContain","Type":"Task","ResultPath":"$.Result.GetMatchTagResource","ResultSelector":{"Arns.$":"$..ResourceTagMappingList[*].ResourceARN"},"Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:resourcegroupstaggingapi:getResources","Parameters":{"ResourceTypeFilters":["codepipeline:pipeline"],"TagFilters":[{"Key.$":"$.params.tagKey","Values.$":"$.params.tagValues"}]}},"CheckTagFilterArnsContain":{"Type":"Pass","ResultPath":"$.Result.TagFilterArnsContain","Parameters":{"Is.$":"States.ArrayContains($.Result.GetMatchTagResource.Arns, $.Result.Pipeline.Arn)"},"Next":"CheckFoundTagMatch"},"CheckFoundTagMatch":{"Type":"Choice","Choices":[{"Variable":"$.Result.TagFilterArnsContain.Is","BooleanEquals":true,"Next":"CheckPipelineStateMatch"}],"Default":"UnMatchPipelineTagFilter"},"UnMatchPipelineTagFilter":{"Type":"Pass","End":true},"CheckPipelineStateMatch":{"Type":"Choice","Choices":[{"Or":[{"Variable":"$.event.detail.state","StringEquals":"STARTED"},{"Variable":"$.event.detail.state","StringEquals":"SUCCEEDED"},{"Variable":"$.event.detail.state","StringEquals":"RESUMED"},{"Variable":"$.event.detail.state","StringEquals":"FAILED"},{"Variable":"$.event.detail.state","StringEquals":"STOPPING"},{"Variable":"$.event.detail.state","StringEquals":"STOPPED"},{"Variable":"$.event.detail.state","StringEquals":"SUPERSEDED"}],"Next":"GeneratePipelineUrl"}],"Default":"UnMatchStatus"},"UnMatchStatus":{"Type":"Pass","End":true},"GeneratePipelineUrl":{"Type":"Pass","ResultPath":"$.Generate.PipelineUrl","Parameters":{"Value.$":"States.Format('https://{}.console.aws.amazon.com/codesuite/codepipeline/pipelines/{}/view?region={}', $.event.region, $.event.detail.pipeline, $.event.region)"},"Next":"GenerateTopicSubject"},"GenerateTopicSubject":{"Type":"Pass","ResultPath":"$.Generate.Topic.Subject","Parameters":{"Value.$":"States.Format('{} [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', States.ArrayGetItem($.Definition.StateEmojis[?(@.name == $.event.detail.state)].emoji, 0), $.event.detail.state, $.event.account, $.event.region)"},"Next":"GeneratedPipelineMessage"},"GeneratedPipelineMessage":{"Type":"Pass","ResultPath":"$.Generate.Topic.Message","Parameters":{"Value.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline : {}\\nState : {}\\nTime : {}\\nURL : {}\\n', $.event.account, $.event.region, $.event.detail.pipeline, $.event.detail.state, $.event.time, $.Generate.PipelineUrl.Value)"},"Next":"SendNotification"},"SendNotification":{"Next":"Succeed","Type":"Task","ResultPath":"$.snsResult","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish","Parameters":{"TopicArn":"",
              {
                "Ref": "NotificationTopicEB7A0DF1",
              },
              "","Message.$":"$.Generate.Topic.Message.Value","Subject.$":"$.Generate.Topic.Subject.Value"}},"Succeed":{"Type":"Succeed"}},"TimeoutSeconds":300}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineEventsRoleDBCDECD1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineEventsRoleDefaultPolicyFB602CA9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "StateMachine2E01A3A5",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineEventsRoleDefaultPolicyFB602CA9",
        "Roles": [
          {
            "Ref": "StateMachineEventsRoleDBCDECD1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "codepipeline:getPipeline",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:codepipeline:*:",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":*",
                  ],
                ],
              },
            },
            {
              "Action": "tag:GetResources",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "NotificationTopicEB7A0DF1",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
